File: writer.go
package shmring

import (
	"context"
	"time"

	"github.com/gofsd/shmring/backend"
)

// Writer is the producer side of a ring buffer. A Writer must only be used
// from a single goroutine at a time.
type Writer struct {
	st       backend.Storage
	capacity int64
	mask     int64
	opt      options

	tail       uint32 // local authoritative copy; persisted to storage on every write
	cachedHead uint32 // last head value observed from storage
	closed     bool
}

// NewWriter initializes a fresh ring buffer header on st and returns the
// producer handle for it. capacity must be a positive power of two, and st
// must be at least headerSize+capacity bytes.
//
// NewWriter is the low-level entry point used by CreateShm; use it directly
// when running the ring buffer over a custom backend.Storage (for example
// backend.MemStorage in tests).
func NewWriter(st backend.Storage, capacity int64, opts ...Option) (*Writer, error) {
	if err := validateCapacity(capacity); err != nil {
		return nil, err
	}
	if st.Size() < headerSize+capacity {
		return nil, ErrStorageTooSmall
	}
	if err := initHeader(st, capacity); err != nil {
		return nil, err
	}
	o := defaultOptions()
	for _, opt := range opts {
		opt(&o)
	}
	return &Writer{
		st:       st,
		capacity: capacity,
		mask:     capacity - 1,
		opt:      o,
	}, nil
}

// TryWrite writes as much of p as currently fits in the ring buffer without
// blocking, returning the number of bytes written. It returns (0, nil) if
// the buffer is full, and (0, ErrClosed) if the Writer has been closed.
func (w *Writer) TryWrite(p []byte) (int, error) {
	if w.closed {
		return 0, ErrClosed
	}
	if len(p) == 0 {
		return 0, nil
	}

	free := w.capacity - int64(w.tail-w.cachedHead)
	if free < int64(len(p)) {
		// The cached head may be stale (the reader has consumed more than
		// we've observed); refresh it before concluding there's no room.
		head, err := readUint32At(w.st, offHead)
		if err != nil {
			return 0, err
		}
		w.cachedHead = head
		free = w.capacity - int64(w.tail-w.cachedHead)
	}
	if free <= 0 {
		return 0, nil
	}

	n := int64(len(p))
	if n > free {
		n = free
	}

	start := int64(w.tail) & w.mask
	if start+n <= w.capacity {
		if _, err := w.st.WriteAt(p[:n], headerSize+start); err != nil {
			return 0, err
		}
	} else {
		first := w.capacity - start
		if _, err := w.st.WriteAt(p[:first], headerSize+start); err != nil {
			return 0, err
		}
		if _, err := w.st.WriteAt(p[first:n], headerSize); err != nil {
			return 0, err
		}
	}

	w.tail += uint32(n)
	if err := writeUint32At(w.st, offTail, w.tail); err != nil {
		return 0, err
	}
	return int(n), nil
}

// Write writes all of p to the ring buffer, blocking until space is
// available. It implements io.Writer. Write only returns a short count
// alongside a non-nil error; on success n == len(p).
func (w *Writer) Write(p []byte) (int, error) {
	return w.WriteContext(context.Background(), p)
}

// WriteContext is like Write but returns ctx.Err() if ctx is done before
// all of p has been written.
func (w *Writer) WriteContext(ctx context.Context, p []byte) (int, error) {
	waiter, canWait := w.st.(backend.WaiterStorage)
	written := 0
	wait := w.opt.minPoll
	for written < len(p) {
		n, err := w.TryWrite(p[written:])
		written += n
		if err != nil {
			return written, err
		}
		if n > 0 {
			wait = w.opt.minPoll
			continue
		}
		if err := ctx.Err(); err != nil {
			return written, err
		}
		if canWait {
			// Block on a real wakeup tied to offHead instead of polling.
			// The wait is capped at maxPoll (when ctx is cancellable) so
			// ctx cancellation is still noticed promptly; that cap never
			// delays a real wakeup, which interrupts Wait immediately
			// regardless of the timeout passed to it.
			waiter.Wait(offHead, w.cachedHead, waitSlice(ctx, w.opt.maxPoll))
			continue
		}
		select {
		case <-ctx.Done():
			return written, ctx.Err()
		case <-time.After(wait):
		}
		if wait *= 2; wait > w.opt.maxPoll {
			wait = w.opt.maxPoll
		}
	}
	return written, nil
}

// waitSlice returns how long a WaiterStorage.Wait call should block for:
// indefinitely if ctx can never be cancelled (context.Background(), the
// common case for the plain Write/Read methods), otherwise capped at cap
// so an already-cancelled or soon-to-expire ctx is still noticed within
// that bound.
func waitSlice(ctx context.Context, cap time.Duration) time.Duration {
	if ctx.Done() == nil {
		return 0 // WaiterStorage.Wait treats <= 0 as "wait indefinitely"
	}
	if dl, ok := ctx.Deadline(); ok {
		if remaining := time.Until(dl); remaining < cap {
			if remaining <= 0 {
				// Deadline already passed; take the shortest real wait
				// rather than 0, which WaiterStorage.Wait reads as
				// "indefinitely". The loop's ctx.Err() check catches this
				// on the next iteration regardless.
				return time.Nanosecond
			}
			return remaining
		}
	}
	return cap
}

// Close marks the ring buffer as closed. Any data already written remains
// available for the Reader to drain; once drained, the Reader's Read calls
// return ErrClosed. Close does not release the underlying storage — call
// Storage.Close (or keep using the Writer's storage via Reader) separately
// once both sides are done, since the Reader may still be reading.
func (w *Writer) Close() error {
	if w.closed {
		return nil
	}
	w.closed = true
	if err := writeUint32At(w.st, offClosed, 1); err != nil {
		return err
	}
	// A Reader blocked in WaiterStorage.Wait watches offTail, not
	// offClosed (there's no data to wait for otherwise), so closing
	// without a final write needs an explicit nudge here. Re-writing
	// tail's current, unchanged value is a deliberate no-op purely for
	// writeUint32At's wake-after-store side effect.
	return writeUint32At(w.st, offTail, w.tail)
}

// CloseStorage closes the underlying storage in addition to marking the
// ring buffer closed. For OS shared memory this unmaps the segment and, on
// the creating side, removes it. Call this once no other process still
// needs the storage.
func (w *Writer) CloseStorage() error {
	if err := w.Close(); err != nil {
		return err
	}
	return w.st.Close()
}

File: rust/src/writer.rs
use std::io;
use std::time::{Duration, Instant};

use crate::backend::Storage;
use crate::error::{Error, Result};
use crate::header::{
    init_header, read_u32_at, validate_capacity, wait_slice, write_u32_at, HEADER_SIZE, OFF_CLOSED,
    OFF_HEAD, OFF_TAIL,
};
use crate::options::Options;

/// The producer side of a ring buffer. A `Writer` must only be used from a
/// single thread at a time.
pub struct Writer<S: Storage> {
    storage: S,
    capacity: u64,
    mask: u64,
    options: Options,

    tail: u32,        // local authoritative copy; persisted to storage on every write
    cached_head: u32, // last head value observed from storage
    closed: bool,
}

impl<S: Storage> Writer<S> {
    /// Initializes a fresh ring buffer header on `storage` and returns the
    /// producer handle for it. `capacity` must be a positive power of two,
    /// and `storage` must be at least `header size + capacity` bytes.
    ///
    /// This is the low-level entry point used by [`create_shm`](crate::create_shm);
    /// use it directly to run the ring buffer over a custom [`Storage`]
    /// (for example [`MemStorage`](crate::backend::MemStorage) in tests).
    pub fn new(storage: S, capacity: u64, options: Options) -> Result<Self> {
        validate_capacity(capacity)?;
        if storage.size() < HEADER_SIZE + capacity {
            return Err(Error::StorageTooSmall);
        }
        init_header(&storage, capacity)?;
        Ok(Writer {
            storage,
            capacity,
            mask: capacity - 1,
            options,
            tail: 0,
            cached_head: 0,
            closed: false,
        })
    }

    /// Writes as much of `buf` as currently fits in the ring buffer without
    /// blocking, returning the number of bytes written. Returns `Ok(0)` if
    /// the buffer is full, and `Err(Error::Closed)` if the writer has been
    /// closed.
    pub fn try_write(&mut self, buf: &[u8]) -> Result<usize> {
        if self.closed {
            return Err(Error::Closed);
        }
        if buf.is_empty() {
            return Ok(0);
        }

        let mut free = self.capacity as i64 - self.tail.wrapping_sub(self.cached_head) as i64;
        if free < buf.len() as i64 {
            // The cached head may be stale (the reader has consumed more
            // than we've observed); refresh it before concluding there's
            // no room.
            self.cached_head = read_u32_at(&self.storage, OFF_HEAD)?;
            free = self.capacity as i64 - self.tail.wrapping_sub(self.cached_head) as i64;
        }
        if free <= 0 {
            return Ok(0);
        }

        let mut n = buf.len() as i64;
        if n > free {
            n = free;
        }
        let n = n as u64;

        let start = self.tail as u64 & self.mask;
        if start + n <= self.capacity {
            self.storage
                .write_at(&buf[..n as usize], HEADER_SIZE + start)?;
        } else {
            let first = self.capacity - start;
            self.storage
                .write_at(&buf[..first as usize], HEADER_SIZE + start)?;
            self.storage
                .write_at(&buf[first as usize..n as usize], HEADER_SIZE)?;
        }

        self.tail = self.tail.wrapping_add(n as u32);
        write_u32_at(&self.storage, OFF_TAIL, self.tail)?;
        Ok(n as usize)
    }

    /// Writes all of `buf`, blocking until space is available or `timeout`
    /// elapses since the call started. On success, the full buffer was
    /// written. On `Err(Error::Timeout)`, a prefix of `buf` may already
    /// have been written (and is already visible to the reader) before the
    /// deadline elapsed; the ring buffer has no way to report exactly how
    /// much, since that data cannot be un-written. Mirrors Go's
    /// `WriteContext`.
    pub fn write_timeout(&mut self, buf: &[u8], timeout: Duration) -> Result<usize> {
        self.write_until(buf, Some(Instant::now() + timeout))
    }

    fn write_until(&mut self, buf: &[u8], deadline: Option<Instant>) -> Result<usize> {
        let can_wait = self.storage.supports_wait();
        let mut written = 0usize;
        let mut wait = self.options.min_poll;
        while written < buf.len() {
            let n = self.try_write(&buf[written..])?;
            written += n;
            if n > 0 {
                wait = self.options.min_poll;
                continue;
            }
            if let Some(deadline) = deadline {
                if Instant::now() >= deadline {
                    return Err(Error::Timeout);
                }
            }
            if can_wait {
                // Block on a real wakeup tied to OFF_HEAD instead of
                // sleeping; see wait_slice for why the wait is bounded
                // when there's a deadline. A real wakeup interrupts this
                // immediately regardless of that bound.
                self.storage
                    .wait_u32_at(OFF_HEAD, self.cached_head, wait_slice(deadline));
                continue;
            }
            std::thread::sleep(wait);
            wait = (wait * 2).min(self.options.max_poll);
        }
        Ok(written)
    }

    /// Marks the ring buffer as closed. Any data already written remains
    /// available for the reader to drain; once drained, the reader's reads
    /// return end-of-stream. `close` does not release the underlying
    /// storage -- call [`close_storage`](Writer::close_storage) instead
    /// once no other process still needs the storage.
    pub fn close(&mut self) -> Result<()> {
        if self.closed {
            return Ok(());
        }
        self.closed = true;
        write_u32_at(&self.storage, OFF_CLOSED, 1)?;
        // A reader blocked in Storage::wait_u32_at watches OFF_TAIL, not
        // OFF_CLOSED (there's no data to wait for otherwise), so closing
        // without a final write needs an explicit nudge here. Re-storing
        // tail's current, unchanged value is a deliberate no-op purely
        // for store_u32_at's wake-on-store side effect.
        write_u32_at(&self.storage, OFF_TAIL, self.tail)
    }

    /// The last head value this `Writer` has observed from `storage`
    /// (refreshed by `try_write` whenever the cached one looked
    /// insufficient). crate-internal: used by
    /// [`wasm_api`](crate::wasm_api) to drive `Atomics.waitAsync` on
    /// `OFF_HEAD` without duplicating the ring buffer's own bookkeeping.
    #[cfg_attr(
        not(all(target_arch = "wasm32", target_os = "unknown")),
        allow(dead_code)
    )]
    pub(crate) fn cached_head(&self) -> u32 {
        self.cached_head
    }

    /// crate-internal accessor to the underlying storage, for the same
    /// reason as [`cached_head`](Writer::cached_head).
    #[cfg_attr(
        not(all(target_arch = "wasm32", target_os = "unknown")),
        allow(dead_code)
    )]
    pub(crate) fn storage(&self) -> &S {
        &self.storage
    }

    /// Marks the ring buffer closed (see [`close`](Writer::close)) and
    /// additionally closes the underlying storage: for OS shared memory
    /// this unmaps the segment and, on the creating side, removes it. Call
    /// this once no other process still needs the storage.
    pub fn close_storage(mut self) -> Result<()> {
        self.close()?;
        self.storage.close()
    }
}

impl<S: Storage> std::fmt::Debug for Writer<S> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Writer")
            .field("capacity", &self.capacity)
            .field("closed", &self.closed)
            .finish()
    }
}

impl<S: Storage> io::Write for Writer<S> {
    /// Blocks until at least one byte can be written, then writes as much
    /// of `buf` as currently fits. Combine with the [`Write`](io::Write)
    /// trait's default `write_all` to block until the whole buffer is
    /// written, unbounded -- the equivalent of Go's blocking `Write`.
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        if buf.is_empty() {
            return Ok(0);
        }
        let can_wait = self.storage.supports_wait();
        let mut wait = self.options.min_poll;
        loop {
            match self.try_write(buf) {
                Ok(0) => {
                    if can_wait {
                        self.storage.wait_u32_at(OFF_HEAD, self.cached_head, None);
                    } else {
                        std::thread::sleep(wait);
                        wait = (wait * 2).min(self.options.max_poll);
                    }
                }
                Ok(n) => return Ok(n),
                Err(Error::Closed) => {
                    return Err(io::Error::new(io::ErrorKind::BrokenPipe, Error::Closed))
                }
                Err(Error::Io(e)) => return Err(e),
                Err(other) => return Err(io::Error::other(other)),
            }
        }
    }

    fn flush(&mut self) -> io::Result<()> {
        Ok(())
    }
}
